src/runtime/cycle_participant.rs

```diff
@@ -0,0 +1,26 @@
+use std::panic::AssertUnwindSafe;
+
+use crate::Cycle;
+
+pub(crate) struct CycleParticipant {
+    cycle: Cycle,
+}
+
+impl CycleParticipant {
+    pub(crate) fn new(cycle: Cycle) -> Self {
+        Self { cycle }
+    }
+
+    pub(crate) fn throw(self) -> ! {
+        std::panic::resume_unwind(Box::new(self));
+    }
+
+    pub(crate) fn recover<T>(execute: impl FnOnce() -> T, recover: impl FnOnce(Cycle) -> T) -> T {
+        std::panic::catch_unwind(AssertUnwindSafe(execute)).unwrap_or_else(|err| {
```

Comment:
Can we, instead of asserting unwind-safeness, just require it by, eg, adding some trait bounds to db?

Reply:
Yeah, thanks for calling that out, I wasn't happy with that part of the PR. I think we probably *do* want `salsa::Database` to just require `UnwindSafe`, and that would suffice. It seems like we're pretty "all in" on unwinding so...